test/Main.hs

```diff
@@ -0,0 +1,207 @@
+stringToLBS = LBS.fromStrict . TE.encodeUtf8 . T.pack
+
+main :: IO ()
+main = do
+  -- Discover the test directory path (relative to the current working directory)
+  cwd <- getCurrentDirectory
+  let testDir = cwd </> "test"
+
+  -- Change to the test directory so that agda2hs can find the agda-lib
+  setCurrentDirectory testDir
+
+  -- Create a temporary build directory
+  tmpDir <- getTemporaryDirectory
+  let buildDir = tmpDir </> "agda2hs-test-build"
+  createDirectoryIfMissing True buildDir
+
+  -- Discover test cases
+  succeedTests <- discoverSucceedTests testDir buildDir
+  failTests <- discoverFailTests testDir buildDir
+
+  -- Run tests
+  defaultMain $ testGroup "agda2hs"
+    [ testGroup "Succeed" succeedTests
+    , testGroup "Fail" failTests
+    ]
+
+-- | Discover all .agda files under the Succeed directory recursively.
+-- Files are ordered by: modification date (newest first), then golden value
+-- modification date (newest first), then alphabetically.
+discoverSucceedTests :: FilePath -> FilePath -> IO [TestTree]
+discoverSucceedTests testDir buildDir = do
+  agdaFiles <- findAgdaFilesRecursive (testDir </> "Succeed")
+  sortedFiles <- sortByModTime agdaFiles (\f -> dropExtension f ++ ".hs")
+  forM sortedFiles $ \agdaFile -> do
+    let testName = dropExtension (makeRelative (testDir </> "Succeed") agdaFile)
+        goldenFile = dropExtension agdaFile ++ ".hs"
+    return $ succeedTest testDir buildDir testName agdaFile goldenFile
+
+-- | Discover all .agda files under the Fail directory.
+-- Files are ordered by: modification date (newest first), then golden value
+-- modification date (newest first), then alphabetically.
+discoverFailTests :: FilePath -> FilePath -> IO [TestTree]
+discoverFailTests testDir buildDir = do
+  agdaFiles <- findAgdaFilesRecursive (testDir </> "Fail")
+  sortedFiles <- sortByModTime agdaFiles (\f -> dropExtension f ++ ".err")
+  forM sortedFiles $ \agdaFile -> do
+    let testName = dropExtension (makeRelative (testDir </> "Fail") agdaFile)
+        goldenFile = dropExtension agdaFile ++ ".err"
+    return $ failTest testDir buildDir testName agdaFile goldenFile
+
+-- | Sort files by modification time (newest first), then by golden value
+-- modification time (if it exists), then alphabetically.
+sortByModTime :: [FilePath] -> (FilePath -> FilePath) -> IO [FilePath]
+sortByModTime files goldenPath = do
+  filesWithTimes <- forM files $ \f -> do
+    agdaTime <- getModificationTime f
+    let golden = goldenPath f
+    goldenExists <- doesFileExist golden
+    goldenTime <- if goldenExists
+                    then Just <$> getModificationTime golden
+                    else return Nothing
+    -- Use Down to sort newest first; standard tuple ordering does the rest
+    return (Down agdaTime, Down goldenTime, f)
+  return $ map (\(_,_,f) -> f) $ sort filesWithTimes
+
+-- | Find all .agda files recursively in a directory.
+findAgdaFilesRecursive :: FilePath -> IO [FilePath]
```

Comment:
This function could just be a call to `findFilesWith` (from `System.Directory`) instead of doing manual recursion.